### PR TITLE
feat: enhanced API for synchronous replication

### LINKED ISSUE #5095: [Feature]: `synchronous_standby_names` configuration
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

This feature addresses limitations in the current synchronous replication implementation in CloudNativePG, specifically highlighted in issue #3969. The primary concerns are:

- **Self-healing over consistency**: Current implementation prioritizes self-healing, which can be counter-intuitive for PostgreSQL DBAs who expect consistency to be paramount.
- **Limited configuration options**: Only quorum-based configurations (`ANY`) within the local cluster are currently supported.
- **Lack of priority-based replication**: Priority-based synchronous replication (`FIRST`) is unavailable.
- **Customisation restrictions**: Users cannot customise the `synchronous_standby_names` content, limiting the use of external clusters as synchronous standbys.

### Describe the solution you'd like

The solution is inspired by the proposal made by @fcanovai [in this comment](https://github.com/cloudnative-pg/cloudnative-pg/issues/3969#issuecomment-2228857499), with modifications for clarity and functionality.

#### Example

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: acdc
spec:
  instances: 3

  storage:
    size: 1G

  postgresql:
    synchronous:
      method: first  # Method can be either 'any' (quorum) or 'first' (priority)
      number: 2  # Number of sync replicas expected
      maxStandbyNamesFromCluster: 1  # Max number of local cluster pods in 'synchronous_standby_names'
      standbyNamesPre:  # User-defined list of application names to add before local ones
        - angus
      standbyNamesPost: []  # User-defined list of application names to add after local ones
```

#### Expected Syntax

The expected `synchronous_standby_names` syntax, as per PostgreSQL documentation, for priority (`FIRST`) and quorum (`ANY`) sync replication is:

```
[FIRST] num_sync ( standby_name [, ...] )
ANY num_sync ( standby_name [, ...] )
```

Using the proposed configuration, the syntax will be constructed as follows:

```
<method> <number> (<standbyNamesPre>, <standbyNamesBody>, <standbyNamesPost>)
```

Where `<standbyNamesBody>` is dynamically built by CloudNativePG based on the HA replicas in the cluster, limited by `maxStandbyNamesFromCluster` (default: `-1`, meaning all expected).

Example output for the above configuration:

```
FIRST 2 (angus, acdc-2)
```

#### Use case: Standard Quorum-based

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: acdc
spec:
  instances: 3

  storage:
    size: 1G

  postgresql:
    synchronous:
      method: any
      number: 1
```

The above will produce this:

```
ANY 1 (acdc-2, acdc-3)
# TODO: check if also acdc-1 should be there
```

#### Use-case: Synchronous replica cluster

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: acdc
spec:
  instances: 3

  storage:
    size: 1G

  postgresql:
    synchronous:
      method: first
      number: 2
      maxStandbyNamesFromCluster: 1
      standbyNamesPre:
      - `malcolm`
```

The above will produce this:

```
FIRST 2 (malcolm, acdc-2)
```

This configuration prioritises a replica cluster in another data centre (`malcolm`) over the second local replica.

#### Differences with the current behaviour

The current CloudNativePG implementation uses `minSyncReplicas` and `maxSyncReplicas`, which will be deprecated by this new feature. Both methods will coexist until the API v2, where the old syntax will be removed. Users will not be allowed to define both `postgresql.synchronous` and `minSyncReplicas`/`maxSyncReplicas` simultaneously to avoid conflicts.

### Describe alternatives you've considered

See #3969 

### Additional context

This implementation will implicitly fix #5000 

### Backport?

No

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by leonardoce**
Introducing an improved API for explicit synchronous replication configuration, supporting quorum-based and priority list strategies.

This update allows full customization of the `synchronous_standby_names` option, providing greater control and flexibility. The focus of this patch is on data durability over high availability.

The previous synchronous replication method is now deprecated but will continue to be supported until further notice, ensuring a smooth transition.

Closes: #5095

### CODE REVIEW

**Comment by leonardoce** [comment]
We miss documentation and E2e tests here.

**Review by armru — APPROVED** [review]

File: pkg/postgres/replication/replication.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+// GetSynchronousStandbyNames gets the value to be applied
+// to synchronous_standby_names
+func GetSynchronousStandbyNames(cluster *apiv1.Cluster) string {
+	if cluster.Spec.PostgresConfiguration.Synchronous != nil {
+		return explicitSynchronousStandbyNames(cluster)
+	}
+
+	return legacySynchronousStandbyNames(cluster)
+}
```
Comment (armru) [inline pkg/postgres/replication/replication.go:41]: I don't have a strong opinion, but it is worth considering having these APIs attached to the cluster as methods, given that they are required to interpret the data. I do understand that the proposed design is trying to encapsulate the logic in a separate package

**Comment by NiccoloFei** [comment]
E2E tests passed [here](https://github.com/cloudnative-pg/cloudnative-pg/pull/5148#issuecomment-2252899140)
Running also on all cloud providers in: https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/10114935645

**Review by NiccoloFei — APPROVED** [review]

**Comment by NiccoloFei** [comment]
> E2E tests passed [here](https://github.com/cloudnative-pg/cloudnative-pg/pull/5148#issuecomment-2252899140) Running also on all cloud providers in: https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/10114935645

Second Round of E2Es result: 
```
🟢 - local: 100.0% success. (0 out of 1975 tests failed)
🟢 - openshift: 99.7% success. (2 out of 798 tests failed)
🟢 - aks: 99.8% success. (1 out of 501 tests failed)
🟢 - gke: 99.4% success. (3 out of 489 tests failed)
🟢 - eks: 99.8% success. (1 out of 640 tests failed)
```
Tests are green besides some casual failures on cloud providers only, all unrelated from this patch, which will be dealt with in future issues.

**Pull request merged**